nifi-nar-bundles/nifi-graph-bundle/nifi-graph-processors/src/main/java/org/apache/nifi/processors/graph/ExecuteGraphQueryRecord.java

```diff
@@ -0,0 +1,289 @@
+        return result.getSelectedFields()
+                .filter(fv -> fv.getValue() != null)
+                .map(FieldValue::getValue)
+                .collect( Collectors.toList());
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        FlowFile input = session.get();
+        if ( input == null ) {
+            return;
+        }
+
+        FlowFile output = session.create(input);
+        FlowFile graph = session.create(input);
+        session.getProvenanceReporter().clone(input, graph);
+
+        String recordScript = context.getProperty(SUBMISSION_SCRIPT)
+                .evaluateAttributeExpressions(input)
+                .getValue();
+
+        boolean failed = false;
+        boolean error = false;
+        final AtomicLong errors = new AtomicLong();
+        long delta = 0;
+        Map<String, String> attributes = new HashMap<>();
+        List<Map<String,Object>> graphResponses = new ArrayList<>();
+        try (InputStream is = session.read(input);
+             OutputStream os = session.write(output);
+             OutputStream graphOutputStream = session.write(graph);
+             RecordReader reader = recordReaderFactory.createRecordReader(attributes, is, -1l, getLogger());
+             RecordSetWriter writer = recordSetWriterFactory.createWriter(getLogger(), reader.getSchema(), os, input.getAttributes());
+        ) {
+            List<PropertyDescriptor> dynamic = context.getProperties()
+                    .keySet().stream()
+                    .filter(PropertyDescriptor::isDynamic)
+                    .collect(Collectors.toList());
+            Record record;
+
+            long start = System.currentTimeMillis();
+            long recordIndex = 0;
+            long batchStart = start;
+            writer.beginRecordSet();
+            while ((record = reader.nextRecord()) != null) {
+                Map<String, Object> dynamicPropertyMap = new HashMap<>();
+                for (PropertyDescriptor entry : dynamic) {
+                        if(!dynamicPropertyMap.containsKey(entry.getName())) {
+                            dynamicPropertyMap.put(entry.getName(), getRecordValue(context, entry, record, input));
+                        }
+                }
+                dynamicPropertyMap.putAll(input.getAttributes());
+                graphResponses.addAll(executeQuery(recordScript, dynamicPropertyMap));
+                recordIndex++;
+                if (recordIndex % 100 == 0) {
+                    long now = System.currentTimeMillis();
+                    batchStart = System.currentTimeMillis();
+                }
+            }
+            try {
+                String graphOutput = mapper.writerWithDefaultPrettyPrinter().writeValueAsString(graphResponses);
+                graphOutputStream.write(graphOutput.getBytes(StandardCharsets.UTF_8));
+                session.getProvenanceReporter().modifyContent(graph);
```

Comment:
Remove this when you make the list-related change so you don't add `List#size()` extraneous provenance events.

Reply:
Removed.